### ISSUE TITLE
Anchor categories are showing products of disabled subcategories

### ISSUE THREAD

**Issue by koenner01** [magento/magento2#9002]
Categories that are set to anchor 'Yes' and have disabled subcategories will show the products from those disabled subcategories on the frontend

### Preconditions
1. PHP7.0 
2. MG2 2.1.x (tested on 2.1.2, 2.1.4 and 2.1.5)

### Steps to reproduce
1. Create a simple product 'Test' that is in stock and visible on catalog
2. Create a category A (don't add products to it) 
3. Create a subcategory B (parent is A) and add our 'Test' product to it
4. Goto category A on frontend

### Expected result
1. Category A should not have any products results 

### Actual result
1. Category A is showing our 'Test' product

I did some digging into this issue and found a post from 2013 about MG1 having this exact same problem (https://www.proxiblue.com.au/blog/Collection_of_products_in_all_child_categories/).

**Comment by smoskaluk** [comment]
Internal ticket MAGETWO-70303 is created

**Comment by markusZW** [comment]
Is there any update on this issue?
In this scenario the "Test"-product should definitly not be visible in "Category A" if "Category B" is disabled... From my point of view this is a serious problem!

**Comment by ghost** [comment]
Is there any update on this issue?
In this scenario the _Test_ product should definitly not be visible in _Category A_ if _Category B_ is disabled... From my point of view this is a serious problem!

**Comment by adavis0000** [comment]
I am investigating this issue also, on EE 1.14. It definitely feels like a bug and not a feature. Disabling a subcategory should certainly break the association of any products in it to the parent category.

**Comment by magento-engcom-team** [comment]
@koenner01, thank you for your report.
We've created internal ticket(s) MAGETWO-70303 to track progress on the issue.

**Comment by danslo** [comment]
@ishakhsuvarov Why was this one moved back to TODO? Seemed it was fixed in #11937.

**Comment by ghost** [comment]
Is there a work around available until this is released?

**Comment by HannesJa** [comment]
This helped us. I modified the code in pull request https://github.com/magento/magento2/pull/14320 so it only modifies the internals of the functions, and doesn't add any parameters to the protected methods. But the idea is exactly the same as @drashmk wrote in the pull request. Here's the diff:

```diff
--- a/vendor/magento/module-catalog/Model/Indexer/Category/Product/AbstractAction.php
+++ b/vendor/magento/module-catalog/Model/Indexer/Category/Product/AbstractAction.php
@@ -109,6 +109,11 @@
     private $queryGenerator;
 
     /**
+     * @var \Magento\Store\Api\Data\StoreInterface
+     */
+    private $currentStore;
+
+    /**
      * @param ResourceConnection $resource
      * @param \Magento\Store\Model\StoreManagerInterface $storeManager
      * @param \Magento\Catalog\Model\Config $config
@@ -144,6 +149,7 @@
     {
         foreach ($this->storeManager->getStores() as $store) {
             if ($this->getPathFromCategoryId($store->getRootCategoryId())) {
+                $this->currentStore = $store;
                 $this->reindexRootCategory($store);
                 $this->reindexAnchorCategories($store);
                 $this->reindexNonAnchorCategories($store);
@@ -552,6 +558,13 @@
      */
     protected function fillTempCategoryTreeIndex($temporaryName)
     {
+        $isActiveAttributeId = $this->config->getAttribute(
+            \Magento\Catalog\Model\Category::ENTITY,
+            'is_active'
+        )->getId();
+        $categoryMetadata = $this->getMetadataPool()->getMetadata(\Magento\Catalog\Api\Data\CategoryInterface::class);
+        $categoryLinkField = $categoryMetadata->getLinkField();
+
         // This finds all children (cc2) that descend from a parent (cc) by path.
         // For example, cc.path may be '1/2', and cc2.path may be '1/2/3/4/5'.
         $temporarySelect = $this->connection->select()->from(
@@ -563,6 +576,20 @@
                 [$this->connection->quoteIdentifier('cc.path'), $this->connection->quote('/%')]
             ),
             ['child_id' => 'entity_id']
+        )->joinInner(
+            ['ccacd' => $this->getTable('catalog_category_entity_int')],
+            'ccacd.' . $categoryLinkField . ' = cc2.' . $categoryLinkField . ' AND ccacd.store_id = 0' .
+            ' AND ccacd.attribute_id = ' . $isActiveAttributeId,
+            []
+        )->joinLeft(
+            ['ccacs' => $this->getTable('catalog_category_entity_int')],
+            'ccacs.' . $categoryLinkField . ' = cc2.' . $categoryLinkField
+            . ' AND ccacs.attribute_id = ccacd.attribute_id AND ccacs.store_id = ' .
+            $this->currentStore->getId(),
+            []
+        )->where(
+            $this->connection->getIfNullSql('ccacs.value', 'ccacd.value') . ' = ?',
+            1
         );
 
         $this->connection->query(
```

**Comment by sdzhepa** [comment]
Hello @koenner01 @chrischerrett @adavis0000 @HannesJa 

I found the issue was fixed 

Will be available for 2.3.1 soon, see commits for more details:
- ea661055fc94c1f5a3c7399463915491acc8c0a5
- 1b8cd089461f20131a8b5e5784238918138e13ae
- 68933c1b128f09b74298f81822879e2bee63d68a


Will be available for 2.2.8 soon, see commits for more details:
- 93e08a710bb5c9c94a62f4a0d7a3d02ad9fef673

This ticket can be closed.
Please, feel free to reopen or create a new one if issue still exists or was not fully fixed

Thank you for feedback and collaboration
